### ISSUE TITLE
Possible release date / project maturity

### ISSUE THREAD

**Issue by greg-is-kub** [rdentato/tst#1]
Hi @rdentato !

I'm currently working with a team on a big project on embedded system. We currently use catch2 as a testing framework which is too heavy and pulls too many dependencies.
We loved the conciseness of your code and how it is really practical and want to integrate it in our project.
Do you think that your library is mature enough to be used as today ? If so could I ask you to do a tag/release of the commit you think is the best ?

Thanks in advance and have a good day !

Greg

**Comment by rdentato** [comment]
Hi @greg-is-kub !
Thanks for your interest, 
Before answering your question I had to be sure it worked with C++ as I only work with C. I've added a test in C++ and it seems to work fine (but it's just one!).
I'm not planning to make any substantial change now that I extended test groups to 8. Just bug fixes and small improvement.
I will verify if check2 has any functionality that you're going to miss to see if I can extend tst to include them.
Also feel free to suggest features you deem appropriate. I will be happy if this small library could be of help for you.

**Comment by rdentato** [comment]
@greg-is-kub tagged and released the version 0.3.0.
The main branch will be stable (only fixes) for a while.
Any major change will be on the `devel` branch.
I believe it is pretty usable (thanks to being very simple),

**Comment by greg-is-kub** [comment]
Well the main feature I think would be really useful is their way to handle SECTIONS where the code will go through each SECTION one by one and start from the beginning of the `TEST_CAS` each time.

```
TEST_CASE("")
{
   int a = 3

   SECTION("1")
    {
        a = 5 
    }
    SECTION("2")
    {    
        CHECK(a == 3); // This test will pass
    }
}
```

However I have no idea on how to implement that & it might require a lot of code which would go againt the minimalist guideline of this repo

**Comment by rdentato** [comment]
If you look at the devel branch, I've added few lines in `tst.h` to allow the following:
```
  tstcase("All sections") {
    int a = -1;

    tstsetup() {
      a = 5;
      tstcheck(a==5);
    }
    tstsection() {
      tstcheck(a==5);
      a = 8;
    }
    tstsection() {
      tstcheck(a==5);
      a = 9;
    }
    tstcleanup() {
      tstcheck(a==5);
      a = 0;
    }
    tstcheck(a==0);
  }
```

All the above checks will pass.

In the pushed version there's a bug I'm ironing.

The `tstsetup` block is executed before each section, the `tstcleanup` section is executed only once at the end of all the sections.
I was initially considering executing it for each section but it really complicates the code.

Would that be good enough?

**Comment by greg-is-kub** [comment]
Thats already really good for now imo, thanks ! 

I got one question about it : So how would you choose when to use `section` rather that `tstcase` ?

**Comment by greg-is-kub** [comment]
I'd like to give you news about the integration of your project

**Comment by greg-is-kub** [comment]
We've talked with our team member about the integration of `tst` in our project.

They are really interested but since the schedule is a bit tight until december so we've decided for now to postpone the integration until the beginning of the year.

**Comment by rdentato** [comment]
If you think it may be of help, keep suggesting features and providing feedback.
This way, if you will want to integrate it into your project, you'll have the features you need and be familiar with them.

**Comment by rdentato** [comment]
I'm changing something wrt the 0.3. I've renamed the current `tstgroup` to `tstif` (execute the tests in the body only if a condition is met), and reused `tstgroup` for the sections:

```
  tstcase("All sections") {
    int a = -1;

    tstgroup("Keeping 5 starting from %d",a) {
      
      tstnote("Performing setup with a = %d",a);
      tstcheck(a==-1);
      a = 5;
      tstcheck(a==5);

      tstsection("Changing to 8 stating from a = %d",a) {
        tstcheck(a==5);
        a = 8;
        tstcheck(a==8);
      }
      tstsection("Changing to 9 stating from a = %d",a) {
        tstcheck(a==5);
        a = 9;
        tstcheck(a==9);
      }

      tstnote("Cleanup from a = %d",a);
      a = -1;
    }

    tstcheck(a==-1);
  }

```

The code at the top of the group serves as setup and the code at the bottom as cleanup.
I'm writing a tutorial (following the example of the check2 one) to explain better the roles of `testcase`, `testif`, `testgroup`, etc.

**Comment by rdentato** [comment]
@greg-is-kub, I've also added data driven tests. You define a `tstdata` and then loop through it with `tstdatafor()`. Do you believe it could be useful?

```
  tstcase("Check some data") {

   // An array of any type is appropriate for testing.
    struct {int n; char *s;} tstdata[] = {
             {123, "pippo"},
             {431, "pluto"},
             { 93, "topolino"}
    };

   // This will loop through the test data. `tstcurdata` is the current element.
    tstdatafor("Verify edge cases") {
      tstnote("Checking <%d,%s>", tstcurdata.n, tstcurdata.s);
      tstcheck( somecheck(tstcurdata.n , tstcurdata.s) );
    }
  }
```

produces:
```
FILE ▷ t_tst03.c "Data"
CASE┬── Check some data :11
DATA├┬─ Verify edge cases :19
NOTE: Checking <123,pippo> :20
PASS│  somecheck( tstcurdata.n , tstcurdata.s) :21
NOTE: Checking <431,pluto> :20
PASS│  somecheck( tstcurdata.n , tstcurdata.s) :21
NOTE: Checking <93,topolino> :20
PASS│  somecheck( tstcurdata.n , tstcurdata.s) :21
    │└─ :19
    ╰── 0 KO | 3 OK | 0 SKIP
RSLT ▷ 0 KO | 3 OK | 0 SKIP
```

**Comment by greg-is-kub** [comment]
1. I love the `tstgroup` you've created It is really clean, better than tstsetup & tstcleanup. Same for `tstif`, thats a better name.

2. On the other hand I don't see the really see the use of `tstdata`. Why using a for loop on my own array. Having to restrict the name of your array is a bit weird.  
Maybe the point of it would be to have better comprehension for nested cases ? In this case why not create a system to nest the test cases instead ?  This would be useful for complex case handling. Tell me if i got the purpose wrong.

**Comment by rdentato** [comment]
Thanks for the feedback.
The purpose of `tstdata` was to make it clearer that the test was about a specific set of data but you're right that it can be achieved with a simple `for`.  I initially introduced it as a variation of the new `tstgroup`/`tstsection` : if a vector `tstdata` was defined in the group, the sections would have been executed against that vector. But then I noticed it was extremely complex to manage it properly and deemed it not worthy so I simplified it in its current form. 

If it's of little use, as it seems, I'll remove it in the next release.

You may have noticed the `TSTCOMPACT` compilation flag to create a log that is much smaller but still readable. The idea is to create a tool that takes that log and creates a richer (and more colorful) report in HTML. Or an XML for better integration with CI pipelines, etc. 
I was thinking of adding a timestamp at the beginning so that one could archive them and produce even richer reports to show, for example, the progression on the number of tests executed with success over the days. 
Not sure it's worth the effort, though,

You mentioned you are working on an embedded project, I'm also thinking of ways of reducing the footprint as much as possible.
I was also wondering if printing on `stderr` is the right way to go in the case of embedded systems (I've no experience in that field)

Please keep sending suggestions my way, I'm sure this will make this small library ten thousands times better.

**Comment by rdentato** [comment]
@greg-is-kub, I 've further simplified and removed `tstgroup` altogether. Now you do it all within `tstcase` 
```C
    int a = -1;

    tstcase("Keeping 5 starting from %d",a) {
      tstcheck(a==-1);
      a = 5;
      tstcheck(a==5);

      tstsection("Changing to 8 stating from a = %d",a) {
        tstcheck(a==5);
        a = 8;
      }
      tstsection("Changing to 9 stating from a = %d",a) {
        tstcheck(a==5);
        a = 9;
      }
      a = -1;
    }

    tstcheck(a==-1);
```

And also `tstdata` is back:
```C
  tstcase("Data as static array") {

    struct {int n; char *s;} tstdata[] = {
             {123, "pippo"},
             {431, "pluto"},
             { 93, "topolino"}
    };

    tstsection("First check") {
      tstnote("Checking <%d,%s>",tstcurdata.n,tstcurdata.s);
      tstcheck(first_check(tstcurdata.n , tstcurdata.s));
    }

    tstsection("Second check") {
      tstnote("Checking <%d,%s>",tstcurdata.n,tstcurdata.s);
      tstcheck(second_check(tstcurdata.n , tstcurdata.s));
    }
  }
```
Sections will be executed once for each line in the array.

Of course, this does not prevent you from writing more complex tests, it's just something that covers simple cases.
